### PR TITLE
table filters (#27)

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
filter or filter[col name] boolean property in table
So you can activate filtering on every column less one if needed or a specific subset of columns

Try it and let me know what you think

### CODE REVIEW

File: gui/src/components/Taipy/PaginatedTable.tsx [hunk 1]
```diff
@@ -318,14 +333,25 @@ const PaginatedTable = (props: TaipyPaginatedTableProps) => {
...
                                         >
                                             {columns[col].dfid === EDIT_COL ? (
-                                                active && editable && tp_onAdd ? (
-                                                    <IconButton onClick={onAddRowClick} size="small" sx={iconInRowSx}>
-                                                        <AddIcon />
-                                                    </IconButton>
-                                                ) : null
+                                                [
+                                                    active && editable && tp_onAdd ? (
+                                                        <Tooltip title="Add a row" key="addARow">
```
Comment (FabienLelaquais) [inline gui/src/components/Taipy/PaginatedTable.tsx:341]: Is there any smart mechanism that would make it possible to retrieve these text strings from somewhere else?
Thinking localization, obviously...

File: gui/src/components/Taipy/TableFilter.tsx [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+    setFilter: (idx: number, fd: FilterDesc, remove?: boolean) => void;
+}
+
+const anchorOrigin = {
+    vertical: "bottom",
+    horizontal: "right",
+} as PopoverOrigin;
+
+const actionsByType = {
+    string: { "==": "equal", contains: "contains", "!=": "not equal" },
```
Comment (FabienLelaquais) [inline gui/src/components/Taipy/TableFilter.tsx]: "equal" -> "equals"?
"not equal" -> "is not", "does not equal"?

File: gui/src/components/Taipy/TableFilter.tsx [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    number: {
```
Comment (FabienLelaquais) [inline gui/src/components/Taipy/TableFilter.tsx:51]: Same here.
May be a short chat with Martin?

File: gui/src/components/Taipy/TableFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+                            setFilter={updateFilter}
+                        />
+                    ))}
+                    <FilterRow
+                        idx={-(filters.length + 1)}
+                        columns={props.columns}
+                        colsOrder={props.colsOrder}
+                        setFilter={updateFilter}
+                    />
+                    <Box sx={buttonBoxSx}>
```
Comment (FabienLelaquais) [inline gui/src/components/Taipy/TableFilter.tsx:350]: Should these button appear only when relevant? Like when there are no filters yet, what happens?

**Review by dinhlongviolin1 — APPROVED** [review]
works like a charm

File: gui/src/components/Taipy/TableFilter.tsx [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+                        <Button endIcon={<ClearIcon />} onClick={onRemove}>{`Reset list (remove applied filter${
```
Comment (FabienLelaquais) [inline gui/src/components/Taipy/TableFilter.tsx]: I couldn't spot the case where there are no filters yet.

File: gui/src/components/Taipy/TableFilter.tsx [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+                            filters.length > 1 ? "s" : ""
+                        }`}</Button>
+                    </Box>
+                </Box>
+            </Popover>
+        </>
+    );
+};
+
+export default TableFilter;
```
Comment (FabienLelaquais) [inline gui/src/components/Taipy/TableFilter.tsx]: Missing LF?

**Review by FabienLelaquais — APPROVED** [review]
Amazing!

File: gui/src/components/Taipy/PaginatedTable.tsx (hunk 1 above)
Comment (FredLL-Avaiga) [reply gui/src/components/Taipy/PaginatedTable.tsx:341]: yes, we could use a localization library 
I guess we would need a global solution :-)

File: gui/src/components/Taipy/TableFilter.tsx (hunk 3 above)
Comment (FredLL-Avaiga) [reply gui/src/components/Taipy/TableFilter.tsx:51]: I'll contact him

File: gui/src/components/Taipy/TableFilter.tsx (hunk 5 above)
Comment (FredLL-Avaiga) [reply gui/src/components/Taipy/TableFilter.tsx]: the aim of this button is to remove all the applied filters: same as deleting all filters and then clicking apply but without removing the filters from the list

File: gui/src/components/Taipy/TableFilter.tsx (hunk 6 above)
Comment (FredLL-Avaiga) [reply gui/src/components/Taipy/TableFilter.tsx]: mypy didn't complain but ok

File: gui/src/components/Taipy/TableFilter.tsx (hunk 4 above)
Comment (FredLL-Avaiga) [reply gui/src/components/Taipy/TableFilter.tsx:350]: don't really like disappearing buttons ...

Comment (FabienLelaquais) [reply gui/src/components/Taipy/TableFilter.tsx:350]: Inactive?

File: gui/src/components/Taipy/TableFilter.tsx (hunk 5 above)
Comment (FabienLelaquais) [reply gui/src/components/Taipy/TableFilter.tsx]: I'll check this out when it runs on my screen...

File: gui/src/components/Taipy/TableFilter.tsx (hunk 4 above)
Comment (FredLL-Avaiga) [reply gui/src/components/Taipy/TableFilter.tsx:350]: OK

**Pull request merged**